src/initialization/MOM_tracer_initialization_from_Z.F90

```diff
@@ -0,0 +1,1573 @@
+  integer :: isd, ied, jsd, jed ! data domain indices
+
+  integer :: i, j, k, kd
+
+  real, dimension(SZI_(G),SZJ_(G),SZK_(G)+1) :: zi
+  real, dimension(:,:), allocatable :: Depth
+  real, allocatable, dimension(:,:,:), target :: tr_z, mask_z
+  real, allocatable, dimension(:), target :: z_edges_in, z_in
+
+  ! Local variables for ALE remapping
+  real, dimension(:), allocatable :: h1, h2, hTarget, deltaE, tmpT1d
+  real, dimension(:), allocatable :: tmpT1dIn
+  real :: zTopOfCell, zBottomOfCell
+  type(regridding_CS) :: regridCS ! Regridding parameters and work arrays
+  type(remapping_CS) :: remapCS ! Remapping parameters and work arrays
+
+  real, dimension(:,:,:), allocatable :: tmp1
+
+  real :: tempAvg, missing_value
+  integer :: nPoints, ans
+  integer :: id_clock_routine, id_clock_ALE
+  logical :: reentrant_x, tripolar_n
+
+  id_clock_routine = cpu_clock_id('(Initialize tracer from Z)', grain=CLOCK_ROUTINE)
+  id_clock_ALE = cpu_clock_id('(Initialize tracer from Z) ALE', grain=CLOCK_LOOP)
+
+  call cpu_clock_begin(id_clock_routine)
+
+  is = G%isc ; ie = G%iec ; js = G%jsc ; je = G%jec ; nz = G%ke
+  isd = G%isd ; ied = G%ied ; jsd = G%jsd ; jed = G%jed
+  isg = G%isg ; ieg = G%ieg ; jsg = G%jsg ; jeg = G%jeg
+
+  call callTree_enter(trim(mod)//"(), MOM_state_initialization.F90")
+
+  call mpp_get_compute_domain(G%domain%mpp_domain,isc,iec,jsc,jec)
+
+  call get_param(PF, mod, "Z_INIT_HOMOGENIZE", homog, &
+                 "If True, then horizontally homogenize the interpolated \n"//&
+                 "initial conditions.", default=.false.)
+  call get_param(PF, mod, "Z_INIT_ALE_REMAPPING", useALE, &
+                 "If True, then remap straight to model coordinate from file.",&
+                 default=.true.)
+  call get_param(PF, mod, "Z_INIT_REMAPPING_SCHEME", remapScheme, &
+                 "The remapping scheme to use if using Z_INIT_ALE_REMAPPING\n"//&
+                 "is True.", default="PLM")
+
+  ! These are model grid properties, but being applied to the data grid for now.
+  ! need to revisit this (mjh)
+  reentrant_x = .false. ;  call get_param(PF, mod, "REENTRANT_X", reentrant_x,default=.true.)
+  tripolar_n = .false. ;  call get_param(PF, mod, "TRIPOLAR_N", tripolar_n, default=.false.)
+
+
+  if (PRESENT(homogenize)) homog=homogenize
+  if (PRESENT(useALEremapping)) useALE=useALEremapping
+  if (PRESENT(remappingScheme)) remapScheme=remappingScheme
+  recnum=1
+  if (PRESENT(src_var_record)) recnum = src_var_record
+  convert=1.0
+  if (PRESENT(src_var_unit_conversion)) convert = src_var_unit_conversion
+  
+
+  call horiz_interp_and_extrap_tracer(src_file, src_var_nam, convert, recnum, &
+       G, tr_z, mask_z, z_in, z_edges_in, missing_value, reentrant_x, tripolar_n, homog)
+
+  kd = size(z_edges_in,1)-1
+  call pass_var(tr_z,G%Domain)
+  call pass_var(mask_z,G%Domain)
+
+! Done with horizontal interpolation.    
+! Now remap to model coordinates
+  if (useALE) then
+    call cpu_clock_begin(id_clock_ALE)
```

Comment:
I suggest also modularizing this block (within the useALE) so that we can call it for T/S. This would unify all the initialization from z code, except the part where we are finding isopycnal positions.